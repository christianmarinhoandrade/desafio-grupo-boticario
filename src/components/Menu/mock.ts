export default [
  {
    title: 'Cadastro Revendedor',
    path: 'revendedor/inclusao',
    name: 'revendedor-inclusao',
    'data-testid': 'menu-item-id'
  },
  {
    title: 'Cadastro Compras',
    path: 'compras/inclusao',
    name: 'compras-inclusao',
    'data-testid': 'menu-item-id'
  }
]
