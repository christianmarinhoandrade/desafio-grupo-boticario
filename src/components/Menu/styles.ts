import styled, { css } from 'styled-components'

import { Menu as MenuAntd } from 'antd'

export const Menu = styled(MenuAntd)`
  ${({ theme }) => css`
    width: 100%;
    background-color: ${theme.secondaryColor};
    border-bottom: 0px;

    .ant-dropdown-menu-item:hover,
    .ant-dropdown-menu-submenu-title:hover {
      background-color: ${theme.secondaryColor};
    }
  `}
`
