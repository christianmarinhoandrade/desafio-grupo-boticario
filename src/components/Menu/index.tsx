import * as S from './styles'

type ItemMenuProps = {
  path: string
  title: string
  name: string
}

type MenuProps = {
  menu: ItemMenuProps[]
  mode?: 'horizontal' | 'inline'
  mobile?: boolean
  signOut?: () => void
  onClick?: (route: string) => void
}

function Menu({
  menu,
  mode = 'horizontal',
  signOut,
  onClick,
  mobile = false,
  ...props
}: MenuProps) {
  const renderMenu = (menus: ItemMenuProps[]) => {
    return menus.map(({ path, title, name, ...props }: ItemMenuProps) => (
      <S.Menu.Item
        onClick={() => {
          if (path && onClick) onClick(path)
        }}
        key={name}
        {...props}
      >
        {title}
      </S.Menu.Item>
    ))
  }

  return (
    <S.Menu mode={mode} {...props}>
      {renderMenu(menu)}
      {mobile && (
        <S.Menu.Item
          key="sair"
          onClick={() => {
            if (signOut) signOut()
          }}
        >
          Sair
        </S.Menu.Item>
      )}
    </S.Menu>
  )
}

export default Menu
