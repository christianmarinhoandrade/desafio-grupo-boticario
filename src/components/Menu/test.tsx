import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import Menu from '.'
import mockItems from './mock'

describe('<Menu />', () => {
  const menu = mockItems

  it('should render menu default', async () => {
    const { container } = render(<Menu menu={menu} data-testid={'menu-id'} />)
    await waitFor(() => {
      expect(screen.getByTestId('menu-id')).toBeInTheDocument()
      expect(screen.queryByText('Sair')).not.toBeInTheDocument()
      expect(screen.getByTestId('menu-id')).toHaveClass('ant-menu-horizontal')
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  it('should render menu with mode inline', async () => {
    render(<Menu menu={menu} data-testid={'menu-id'} mode="inline" />)
    await waitFor(() => {
      expect(screen.getByTestId('menu-id')).toBeInTheDocument()
      expect(screen.getByTestId('menu-id')).toHaveClass('ant-menu-inline')
    })
  })

  it('should render menu with mobile true', async () => {
    render(<Menu menu={menu} data-testid={'menu-id'} mobile={true} />)
    await waitFor(() => {
      expect(screen.getByTestId('menu-id')).toBeInTheDocument()
      expect(screen.getByText('Sair')).toBeInTheDocument()
    })
  })

  it('should render menu with a two items', async () => {
    render(<Menu menu={menu} data-testid={'menu-id'} mobile={true} />)
    await waitFor(() => {
      expect(screen.getAllByTestId('menu-item-id')).toHaveLength(2)
    })
  })

  it('should render menu to call the methods signOut and onClick if clicked', async () => {
    const onClick = jest.fn()
    const signOut = jest.fn()
    render(
      <Menu
        menu={menu}
        data-testid={'menu-id'}
        onClick={onClick}
        signOut={signOut}
        mobile={true}
      />
    )

    const menuItem = screen.getByText(menu[0].title)
    const menuItemExit = screen.getByText('Sair')

    userEvent.click(menuItemExit)
    userEvent.click(menuItem)
    await waitFor(() => {
      expect(onClick).toHaveBeenCalledTimes(1)
      expect(signOut).toHaveBeenCalledTimes(1)
    })
  })
})
