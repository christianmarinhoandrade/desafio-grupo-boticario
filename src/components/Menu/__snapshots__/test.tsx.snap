// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Menu /> should render menu default 1`] = `
.c0 {
  width: 100%;
  border-bottom: 0px;
}

<ul
  class="ant-menu-overflow ant-menu ant-menu-root ant-menu-horizontal ant-menu-light c0"
  data-menu-list="true"
  data-testid="menu-id"
  role="menu"
  tabindex="0"
>
  <li
    class="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child"
    data-menu-id="rc-menu-uuid-test-revendedor-inclusao"
    data-testid="menu-item-id"
    role="menuitem"
    style="opacity: 1; order: 0;"
    tabindex="-1"
  >
    <span
      class="ant-menu-title-content"
    >
      Cadastro Revendedor
    </span>
  </li>
  <li
    class="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child"
    data-testid="menu-item-id"
    role="menuitem"
    style="opacity: 1; order: 1;"
    tabindex="-1"
  >
    <span
      class="ant-menu-title-content"
    >
      Cadastro Compras
    </span>
  </li>
  <li
    aria-hidden="true"
    class="ant-menu-overflow-item ant-menu-overflow-item-rest ant-menu-submenu ant-menu-submenu-horizontal"
    role="none"
    style="opacity: 0; height: 0px; overflow-y: hidden; order: 9007199254740991; pointer-events: none; position: absolute;"
  >
    <div
      aria-controls="rc-menu-uuid-test-rc-menu-more-popup"
      aria-expanded="false"
      aria-haspopup="true"
      class="ant-menu-submenu-title"
      data-menu-id="rc-menu-uuid-test-rc-menu-more"
      role="menuitem"
      tabindex="-1"
    >
      <span
        aria-label="ellipsis"
        class="anticon anticon-ellipsis"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="ellipsis"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
          />
        </svg>
      </span>
      <i
        class="ant-menu-submenu-arrow"
      />
    </div>
  </li>
</ul>
`
